### PR TITLE
#77/game timer limits

### PR DESCRIPTION
**Pull request by Q16solver**

### CODE REVIEW

File: Assets/Infrastructure/Client/ClientRaceState.cs [hunk 1]
```diff
@@ -14,9 +15,11 @@ public class ClientRaceState : LocalState
     {
         [SerializeField] ServerRaceState serverRaceState;
         [SerializeField] UIView raceView;
+        [SerializeField] int countdownTime = 10;
```
Comment (RFSurdsman) [inline Assets/Infrastructure/Client/ClientRaceState.cs]: @Mysterise Ideas on better variable name? This variable means "show the countdown timer to the user when the race length is less than X".

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 2]
```diff
@@ -14,13 +15,19 @@ namespace Racerr.Infrastructure.Server
     /// </summary>
     public class ServerRaceState : RaceSessionState
     {
+        [SerializeField] int maxRaceLength = 90;
+        [SerializeField] int remainingRaceLengthOnPlayerFinish = 30;
+        [SyncVar] double raceFinishTime;
+        public double LeftOverTime => raceFinishTime - NetworkTime.time;
+
```
Comment (RFSurdsman) [inline Assets/Infrastructure/Server/ServerRaceState.cs]: @Mysterise Ideas on better variable names? especially line 19

File: Assets/Infrastructure/Client/ClientRaceState.cs [hunk 3]
```diff
@@ -62,6 +65,17 @@ protected override void FixedUpdate()
                 // Race Timer. TODO: Extract Race Timer to its own script
                 raceTimerTMP.text = serverRaceState.CurrentRaceLength.ToRaceTimeFormat();
 
+                // Countdown Timer. If any player has finished and reduces the current countdown timer, use that as leftover time.
```
Comment (RFSurdsman) [inline Assets/Infrastructure/Client/ClientRaceState.cs]: Update the comment

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 4]
```diff
@@ -76,7 +83,12 @@ public void NotifyPlayerPassedThroughCheckpoint(Player player, GameObject checkp
         [Server]
         protected override void FixedUpdate()
```
Comment (RFSurdsman) [inline Assets/Infrastructure/Server/ServerRaceState.cs]: Update the summary of the function to explain the new behaviour

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 5]
```diff
@@ -76,7 +83,12 @@ public void NotifyPlayerPassedThroughCheckpoint(Player player, GameObject checkp
...
         {
-            bool isRaceFinished = raceSessionData.FinishedPlayers.Count + raceSessionData.DeadPlayers.Count == raceSessionData.PlayersInRace.Count;
+            if (raceSessionData.FinishedPlayers.Any())
+            {
+                raceFinishTime = Math.Min(raceFinishTime, NetworkTime.time + remainingRaceLengthOnPlayerFinish);
+            }
```
Comment (Mysterise) [inline Assets/Infrastructure/Server/ServerRaceState.cs]: Bring this into another function, FixedUpdate() shouldn't contain logic internals. 

I think the below code should also be modularised.

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 6]
```diff
@@ -14,13 +15,19 @@ namespace Racerr.Infrastructure.Server
     /// </summary>
     public class ServerRaceState : RaceSessionState
     {
+        [SerializeField] int maxRaceLength = 90;
+        [SerializeField] int remainingRaceLengthOnPlayerFinish = 30;
```
[the remark this answers is not in the document]
Comment (Q16solver) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Yes, done

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 7]
```diff
@@ -14,13 +15,19 @@ namespace Racerr.Infrastructure.Server
...
+        [SyncVar] double raceFinishTime;
+        public double LeftOverTime => raceFinishTime - NetworkTime.time;
```
[the remark this answers is not in the document]
Comment (Q16solver) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Yes, done

File: Assets/Infrastructure/Server/ServerRaceState.cs (hunk 5 above)
Comment (Q16solver) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Yes, done

File: Assets/Infrastructure/Client/ClientRaceState.cs (hunk 1 above)
Comment (Mysterise) [reply Assets/Infrastructure/Client/ClientRaceState.cs]: countdownTimeThreshold

File: Assets/Infrastructure/Client/ClientRaceState.cs (hunk 3 above)
Comment (RFSurdsman) [reply Assets/Infrastructure/Client/ClientRaceState.cs]: bump!!

File: Assets/Infrastructure/Server/ServerRaceState.cs (hunk 4 above)
Comment (RFSurdsman) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: bump!!

File: Assets/Infrastructure/Client/ClientRaceState.cs (hunk 1 above)
Comment (Q16solver) [reply Assets/Infrastructure/Client/ClientRaceState.cs]: Yes, done

File: Assets/Infrastructure/Client/ClientRaceState.cs (hunk 3 above)
Comment (Q16solver) [reply Assets/Infrastructure/Client/ClientRaceState.cs]: Yes, done

File: Assets/Infrastructure/Server/ServerRaceState.cs (hunk 4 above)
Comment (Q16solver) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Yes, done

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 8]
```diff
@@ -123,6 +133,15 @@ public void TransitionToIdle()
         {
             ServerStateMachine.Singleton.ChangeState(StateEnum.ServerIdle);
         }
+
+        [Server]
+        void OnAnyPlayerFinished()
```
Comment (Mysterise) [inline Assets/Infrastructure/Server/ServerRaceState.cs]: 1. Missing documentation
2. The current function name is not very good. What does it actually do when OnAnyPlayerFinished()?

Comment (Q16solver) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Yes, done

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 9]
```diff
@@ -123,6 +133,18 @@ public void TransitionToIdle()
         {
             ServerStateMachine.Singleton.ChangeState(StateEnum.ServerIdle);
         }
+
+        /// <summary>
+        /// Changes remaining race time if there are finished players.
```
Comment (Mysterise) [inline Assets/Infrastructure/Server/ServerRaceState.cs]: How does it change the remaining race time?

File: Assets/Infrastructure/Server/ServerRaceState.cs [hunk 10]
```diff
@@ -123,6 +133,19 @@ public void TransitionToIdle()
         {
             ServerStateMachine.Singleton.ChangeState(StateEnum.ServerIdle);
         }
+
+        /// <summary>
+        /// Changes remaining race time to a given time (from the Unity Editor) if there are finished players and the given time 
+        /// is less than the remaining race time.
```
Comment (Mysterise) [inline Assets/Infrastructure/Server/ServerRaceState.cs]: 1. What is the "given time" specifically? Why do we even do this?
2. Don't need to specify (from the Unity Editor)

Comment (Mysterise) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: 3. Why "if there are finished players and the given time is less than the remaining race time."?

Comment (Mysterise) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: I want to drill into you the idea that we don't just have these documentation for the sake of having documentation.

Documentation is to explain semantics that can't just be conveyed easily through code. You should be able to translate what the function is doing in a humanly-understandable way that's not just a 1:1 rephrasing of the code.

Comment (Q16solver) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Added explanation for summary

Comment (Mysterise) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Summary explanation is very hard to understand currently, see if you can make your points clearer.

Comment (RFSurdsman) [reply Assets/Infrastructure/Server/ServerRaceState.cs]: Fixed!

**Pull request merged**